File: lares/lararium_mcp/resolver.py
"""`lar:` URI resolution for the Lararium MCP carrier spine.

Resolution policy follows the current MCP roadmap:
- all-caps roots such as `lar:///AGENTS`, `lar:///LARES`, and
  `lar:///INDEXES/**` live adjacent to tuple roots;
- `AGENTS` and `LARES` map to all-caps files under `lares/`;
- `INDEXES/**` and other all-caps roots may act as virtual resource roots;
- stable tuple root `ha.ka.ba` maps to `lares/ha-ka-ba/`;
- other `*.*.*` tuple roots map to `lares/chapel-perilous-opens/<root>/`.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from urllib.parse import unquote, urlparse


LARES_ROOT = Path(__file__).resolve().parents[1]
STABLE_TUPLE_ROOT = "ha.ka.ba"
CAPS_FILE_ROOTS = {"AGENTS", "LARES"}
VIRTUAL_CAPS_ROOTS = {"INDEXES"}


@dataclass(frozen=True)
class LarResolution:
    uri: str
    root: str
    child_path: tuple[str, ...]
    path: Path | None
    resource_path: str
    kind: str
    exists: bool
    virtual: bool = False

    def to_dict(self) -> dict[str, object]:
        return {
            "uri": self.uri,
            "root": self.root,
            "child_path": list(self.child_path),
            "path": str(self.path) if self.path else None,
            "resource_path": self.resource_path,
            "kind": self.kind,
            "exists": self.exists,
            "virtual": self.virtual,
        }


def _split_lar_uri(uri: str) -> tuple[str, tuple[str, ...]]:
    parsed = urlparse(uri)
    if parsed.scheme != "lar":
        raise ValueError(f"expected lar URI, got {uri!r}")
    if parsed.netloc:
        raise ValueError(f"expected triple-slash lar URI, got {uri!r}")
    parts = tuple(part for part in unquote(parsed.path).lstrip("/").split("/") if part)
    if not parts:
        raise ValueError("lar URI needs a root segment")
    return parts[0], parts[1:]


def _with_md_suffix(path: Path) -> Path:
    if path.suffix:
        return path
    return path.with_suffix(".md")


def _is_tuple_root(root: str) -> bool:
    parts = root.split(".")
    return len(parts) == 3 and all(parts)


def _is_caps_root(root: str) -> bool:
    return root == root.upper() and any(ch.isalpha() for ch in root)


def resolve_lar_uri(uri: str) -> LarResolution:
    """Resolve a `lar:///...` URI into either a repo path or virtual namespace."""

    root, child_path = _split_lar_uri(uri)
    resource_path = "/".join((root, *child_path))

    if root in CAPS_FILE_ROOTS and not child_path:
        path = LARES_ROOT / f"{root}.md"
        return LarResolution(uri, root, child_path, path, resource_path, "caps-file", path.exists())

    if root in VIRTUAL_CAPS_ROOTS or _is_caps_root(root):
        # All-caps roots form a first-class namespace. Some map to files, some
        # remain virtual resources produced by the compiler/index layer.
        return LarResolution(uri, root, child_path, None, resource_path, "caps-virtual", False, True)

    if _is_tuple_root(root):
        if root == STABLE_TUPLE_ROOT:
            base = LARES_ROOT / root.replace(".", "-")
        else:
            base = LARES_ROOT / "chapel-perilous-opens" / root
        path = _with_md_suffix(base.joinpath(*child_path)) if child_path else _with_md_suffix(base)
        return LarResolution(uri, root, child_path, path, resource_path, "tuple-file", path.exists())

    raise ValueError(f"unsupported lar root {root!r} in {uri!r}")


def read_lar_resource(uri: str) -> str:
    """Read a file-backed Lararium resource.

    Virtual roots such as `lar:///INDEXES/**` get materialized by higher index
    layers, not by this raw file reader.
    """

    resolution = resolve_lar_uri(uri)
    if resolution.virtual:
        raise FileNotFoundError(f"{uri} names a virtual lar resource")
    if resolution.path is None or not resolution.path.exists():
        raise FileNotFoundError(f"{uri} does not resolve to an existing file")
    return resolution.path.read_text(encoding="utf-8")

File: packages/lararium-core/src/resolver.ts
/**
 * `lar:` URI resolution for the Lararium carrier spine.
 *
 * Resolution policy:
 * - ha.ka.ba/@lares/{path} → packages/lares/{path}.md  (primary lares corpus path)
 * - ha.ka.ba/@lararium/{pkg}/v{ver}/{path} → packages/{pkg-slug}/memes/{path}.md  (engine corpus)
 * - ha.ka.ba/{other} → packages/lares-chapel-perilous-opens/memes/ha-ka-ba/{other}.md  (legacy compat — remove after URI sweep)
 * - AGENTS, LARES, README → packages/lares/{ROOT}.md  (caps-file legacy alias)
 * - INDEXES/** and other ALL-CAPS roots → virtual namespace (caps-virtual)
 * - other three-segment tuple roots → packages/lares-chapel-perilous-opens/memes/{root}/{path}.md
 *
 * NOTE: `lararium-core` is isomorphic — no `fs`, `path`, or `process` imports.
 * File existence and reading are delegated to the host (lararium-node).
 */

export interface LarResolution {
  readonly uri: string;
  readonly root: string;
  readonly childPath: readonly string[];
  /** Relative path from packages/lares/ root (no leading slash). Null for virtual or engine-corpus roots. */
  readonly resourcePath: string;
  /** Relative path within packages/lares/memes/ — null for virtual, engine-corpus, or chapel roots. Caller resolves against join(laresRoot, "memes"). */
  readonly laresRelPath: string | null;
  /** Relative path within packages/lares-chapel-perilous-opens/memes/ — non-null only for unstable three-segment tuple roots. */
  readonly chapelRelPath: string | null;
  /** Relative path within packages/{pkg-slug}/memes/ — non-null only for engine corpus URIs (@lararium/* scope). */
  readonly engineRelPath: string | null;
  readonly kind: "caps-file" | "caps-virtual" | "tuple-file";
  readonly virtual: boolean;
}

/**
 * Parsed hostful lar authority: `lar://alias:tier@host/...`
 * Trust tier is separate from identity — the host speaks, not overrides.
 */
export interface LarAuthority {
  readonly alias: string;
  readonly tier: string;
  readonly host: string;
}

export interface LarHostfulResolution extends LarResolution {
  readonly authority: LarAuthority;
  /** Hostful records never resolve to lares/ files — they function as exchange records. */
  readonly kind: "caps-virtual";
  readonly virtual: true;
}

// Schema: lar:///ha.ka.ba/@lares/api/v0.1/lararium/lar-uri/uri-roots
const CAPS_FILE_ROOTS = new Set(["AGENTS", "LARES", "README"]);
const VIRTUAL_CAPS_ROOTS = new Set(["INDEXES"]);
const STABLE_TUPLE_ROOT = "ha.ka.ba";
const LARES_SCOPE   = "@lares";
const ENGINE_SCOPE  = "@lararium";

function splitLarUri(uri: string): { root: string; childPath: string[] } {
  const url = new URL(uri);
  if (url.protocol !== "lar:") throw new Error(`expected lar URI, got ${uri}`);
  if (url.host) throw new Error(`expected triple-slash lar URI (hostless), got ${uri} — use parseHostfulLarUri for hostful`);
  const rawPath = decodeURIComponent(url.pathname);
  const parts = rawPath.replace(/^\/+/, "").split("/").filter(Boolean);
  if (parts.length === 0) throw new Error(`lar URI needs a root segment: ${uri}`);
  const [root, ...childPath] = parts as [string, ...string[]];
  return { root, childPath };
}

/**
 * Parse a hostful `lar://alias:tier@host/path` URI.
 * Returns the authority components and a virtual resolution.
 * Hostful records carry lower trust than hostless invariant memes.
 */
export function parseHostfulLarUri(uri: string): LarHostfulResolution {
  const url = new URL(uri);
  if (url.protocol !== "lar:") throw new Error(`expected lar URI, got ${uri}`);
  if (!url.host) throw new Error(`expected hostful lar URI (lar://alias:tier@host/...), got ${uri}`);

  // URL parser splits "alias:tier@host" as username=alias, password=tier, hostname=host
  const alias = decodeURIComponent(url.username);
  const tier = decodeURIComponent(url.password);
  const host = url.hostname;

  const rawPath = decodeURIComponent(url.pathname);
  const parts = rawPath.replace(/^\/+/, "").split("/").filter(Boolean);
  const [root = "", ...childPath] = parts;
  const resourcePath = [root, ...childPath].join("/");

  return Object.freeze({
    uri,
    root,
    childPath: Object.freeze(childPath),
    resourcePath,
    laresRelPath: null,
    chapelRelPath: null,
    engineRelPath: null,
    kind: "caps-virtual" as const,
    virtual: true as const,
    authority: Object.freeze({ alias, tier, host }),
  });
}

/**
 * Returns true if the URI qualifies as a hostful live exchange record.
 * Hostful records must not silently override hostless invariant memes.
 */
export function isHostfulLarUri(uri: string): boolean {
  try {
    const url = new URL(uri);
    return url.protocol === "lar:" && url.host.length > 0;
  } catch {
    return false;
  }
}

function isTupleRoot(root: string): boolean {
  const parts = root.split(".");
  return parts.length === 3 && parts.every((p) => p.length > 0);
}

function isCapsRoot(root: string): boolean {
  return root === root.toUpperCase() && /[A-Z]/.test(root);
}

function withMdSuffix(p: string): string {
  const lastSegment = p.slice(p.lastIndexOf("/") + 1);
  return lastSegment.includes(".") ? p : p + ".md";
}

/**
 * Resolve a `lar:///...` URI into a LarResolution.
 * Does not perform any I/O — existence checking is the caller's responsibility.
 */
export function resolveLarUri(uri: string): LarResolution {
  const { root, childPath } = splitLarUri(uri);
  const resourcePath = [root, ...childPath].join("/");

  // Legacy caps-file roots: lar:///AGENTS → packages/lares/AGENTS.md
  // Kept as alias; canonical form is lar:///ha.ka.ba/@lares/AGENTS
  if (CAPS_FILE_ROOTS.has(root) && childPath.length === 0) {
    const laresRelPath = `${root}.md`;
    return { uri, root, childPath, resourcePath, laresRelPath, chapelRelPath: null, engineRelPath: null, kind: "caps-file", virtual: false };
  }

  if (VIRTUAL_CAPS_ROOTS.has(root) || isCapsRoot(root)) {
    return { uri, root, childPath, resourcePath, laresRelPath: null, chapelRelPath: null, engineRelPath: null, kind: "caps-virtual", virtual: true };
  }

  if (isTupleRoot(root) && root === STABLE_TUPLE_ROOT) {
    // lar:///ha.ka.ba/@lares/{rest} → packages/lares/memes/{rest}.md  (canonical lares path)
    if (childPath[0] === LARES_SCOPE) {
      const rest = childPath.slice(1);
      if (rest.length === 1 && CAPS_FILE_ROOTS.has(rest[0]!)) {
        return { uri, root, childPath, resourcePath, laresRelPath: `${rest[0]}.md`, chapelRelPath: null, engineRelPath: null, kind: "caps-file", virtual: false };
      }
      const joined = rest.length > 0 ? rest.join("/") : "";
      const laresRelPath = joined ? withMdSuffix(joined) : "index.md";
      return { uri, root, childPath, resourcePath, laresRelPath, chapelRelPath: null, engineRelPath: null, kind: "tuple-file", virtual: false };
    }

    // lar:///ha.ka.ba/@lararium/{pkg}/v{ver}/{path} → packages/{pkg-slug}/memes/{path}.md  (engine corpus)
    if (childPath[0] === ENGINE_SCOPE || childPath[0]?.startsWith(ENGINE_SCOPE + "/")) {
      const scopedName = childPath[0] === ENGINE_SCOPE
        ? `${ENGINE_SCOPE}/${childPath[1] ?? ""}`
        : childPath[0]!;
      const afterScope = childPath[0] === ENGINE_SCOPE ? childPath.slice(2) : childPath.slice(1);
      const [_ver, ...pathParts] = afterScope;
      const pkgSlug = scopedName.replace(/^@lararium\//, "lararium-");
      const filePath = pathParts.length > 0 ? pathParts.join("/") : "index";
      const engineRelPath = withMdSuffix(`${pkgSlug}/memes/${filePath}`);
      return { uri, root, childPath, resourcePath, laresRelPath: null, chapelRelPath: null, engineRelPath, kind: "tuple-file", virtual: false };
    }

    // lar:///ha.ka.ba/@{root-doc}[/@{child-doc}][/{path}] — named doc oracle URI.
    //
    // URI grammar law (pos 0-indexed after lar:///):
    //   pos 1  @name  = root doc identity  (one of six reserved roots)
    //   pos 2  @name  = child doc under root  e.g. @catalog/@elyncia corpus
    //   pos 2+ plain  = leaf path under root  e.g. @lararium/rooms/altar-fire
    //   pos 3+ always plain, never @-prefixed
    //
    // Any @-prefixed segment at pos 1 that is not @lares or @lararium resolves
    // as virtual (doc identity, not a file path).
    // @catalog/@{slug} also resolves as virtual (corpus child-doc identity).
    if (childPath[0]?.startsWith("@")) {
      return { uri, root, childPath, resourcePath, laresRelPath: null, chapelRelPath: null, engineRelPath: null, kind: "caps-virtual", virtual: true };
    }

    // Legacy: lar:///ha.ka.ba/{rest} (no scope) — kept in chapel under ha-ka-ba/ subdir. Remove after URI sweep is complete.
    const base = root.replace(/\./g, "-");
    const joined = childPath.length > 0 ? `${base}/${childPath.join("/")}` : base;
    const chapelRelPath = withMdSuffix(joined);
    return { uri, root, childPath, resourcePath, laresRelPath: null, chapelRelPath, engineRelPath: null, kind: "tuple-file", virtual: false };
  }

  // lar:///{a.b.c}/{path} → packages/lares-chapel-perilous-opens/memes/{a.b.c}/{path}.md
  if (isTupleRoot(root)) {
    const joined = childPath.length > 0 ? `${root}/${childPath.join("/")}` : root;
    const chapelRelPath = withMdSuffix(joined);
    return { uri, root, childPath, resourcePath, laresRelPath: null, chapelRelPath, engineRelPath: null, kind: "tuple-file", virtual: false };
  }

  // Adjacent tagspace dirs — legacy; absorb into @lares scope after URI sweep
  if (root === "grammars" || root === "lararium-node") {
    const joined = childPath.length > 0 ? `${root}/${childPath.join("/")}` : root;
    const laresRelPath = withMdSuffix(joined);
    return { uri, root, childPath, resourcePath, laresRelPath, chapelRelPath: null, engineRelPath: null, kind: "tuple-file", virtual: false };
  }

  throw new Error(`unsupported lar root "${root}" in ${uri}`);
}
